### PR TITLE
Quotecall wrapper

### PR DESCRIPTION
**Pull request by cwang25**
Same with paycall, but it's quotecall wrapper

(rebased on top of `hans/paycall-in-builder`)

### CODE REVIEW

File: src/builder/QuotecallWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity ^0.8.23;
+
+import {CodeJarHelper} from "./CodeJarHelper.sol";
+import {IQuarkWallet} from "quark-core/src/interfaces/IQuarkWallet.sol";
+import {Quotecall} from "../Quotecall.sol";
+
+// Helper library to wrap a QuarkOperation from Actions.sol for a Paycall
+library QuotecallWrapper {
+    function wrap(IQuarkWallet.QuarkOperation memory operation, uint256 chainId, uint256 quotedAmount)
```
Comment (kevincheng96) [inline src/builder/QuotecallWrapper.sol:10]: Same comment as in #12. `chainId` can be removed.

File: test/Quotecall.t.sol [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    address constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+
+    function setUp() public {
+        factory = new QuarkWalletProxyFactory(address(new QuarkWallet(new CodeJar(), new QuarkStateManager())));
+        codeJar = QuarkWallet(payable(factory.walletImplementation())).codeJar();
+    }
+
+    /* ===== general tests ===== */
+    function testSimpleTransferTransferAndWrapForQuotecall() public {
```
Comment (kevincheng96) [inline test/Quotecall.t.sol:32]: nit: duplicate `Transfer`?

File: src/builder/QuotecallWrapper.sol [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        internal
+        pure
+        returns (IQuarkWallet.QuarkOperation memory)
+    {
+        bytes[] memory scriptSources = new bytes[](1);
+        scriptSources[0] = type(Quotecall).creationCode;
```
Comment (kevincheng96) [inline src/builder/QuotecallWrapper.sol:16]: Same comment as in #12 about script sources.

File: src/builder/QuotecallWrapper.sol (hunk 1 above)
Comment (cwang25) [reply src/builder/QuotecallWrapper.sol:10]: I think wrap() probably will still need chainId now as we need that to find the right address for pricefeed and payment token for compute the address for quotecall.

File: src/builder/QuotecallWrapper.sol (hunk 3 above)
Comment (cwang25) [reply src/builder/QuotecallWrapper.sol:16]: 👍 Change this PR#14 to rebase on top of #12 which now has the information needed to calculate the address for paycall/quotecall.

File: test/Quotecall.t.sol (hunk 2 above)
Comment (cwang25) [reply test/Quotecall.t.sol]: oops removed :) Thanks!

File: test/QuotecallWrapper.t.sol [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        // Wrap with Quotecall wrapper
+        IQuarkWallet.QuarkOperation memory wrappedQuotecallOp = QuotecallWrapper.wrap(op, 1, "USDC", 20e6);
+
+        // Check the transfer action is wrapped in a Quotecall
+        assertEq(wrappedQuotecallOp.nonce, op.nonce, "nonce should be the same");
+        assertEq(
+            wrappedQuotecallOp.scriptAddress,
+            CodeJarHelper.getCodeAddress(
+                1, abi.encodePacked(type(Quotecall).creationCode, abi.encode(ETH_USD_PRICE_FEED, USDC))
```
Comment (kevincheng96) [inline test/QuotecallWrapper.t.sol:54]: If you rebase on main, `getCodeAddress` no longer takes in a chain id.

**Review by kevincheng96 — APPROVED** [review]
Nice, think this just needs a rebase and everything is good to go

File: test/QuotecallWrapper.t.sol (hunk 4 above)
Comment (cwang25) [reply test/QuotecallWrapper.t.sol]: Ok 👍

**Pull request merged**